### PR TITLE
cleanup(userspace/libscap): avoid the usage of non-portable (glibc specific) `__gnu_cxx::stdio_filebuf`

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind cleanup

**Any specific area of the project related to this PR?**

/area libscap-engine-gvisor

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

Using the glibc specific `__gnu_cxx::stdio_filebuf` breaks while trying to build Falco/libs with zig compiler (see https://github.com/falcosecurity/falco/pull/3307).
In this specific case, we can make use of POSIX standard `popen`.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

There are a couple more usages of `__gnu_cxx::stdio_filebuf`; both are inside libsinsp_e2e tests framework, but we don't really need to build libs tests when we build Falco (with zig) therefore i am skipping them for now.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by FedeDP** [comment]
@LucaGuerra tested with this small test file: 
```C++
#include <sys/wait.h>
#include <ext/stdio_filebuf.h>
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <string>
#include <vector>

struct result {
   int error = 0;
   std::vector<std::string> output;
};

result runsc(char *argv[])
{
	result res;
	int pipefds[2];

	int ret = pipe(pipefds);
	if(ret)
	{
		return res;
	}

	pid_t pid = vfork();
	if(pid > 0)
	{
		int status;

		close(pipefds[1]);
		wait(&status);
		if(!WIFEXITED(status) || WEXITSTATUS(status) != 0)
		{
			res.error = status;
			return res;
		}

		__gnu_cxx::stdio_filebuf<char> filebuf(pipefds[0], std::ios::in);
		std::string line;
		std::istream is(&filebuf);

		while(std::getline(is, line))
		{
			res.output.emplace_back(std::string(line));
		}
	}
	else
	{
		close(pipefds[0]);
		dup2(pipefds[1], STDOUT_FILENO);
		execvp("docker", argv);
		exit(1);
	}

	return res;
}

result runsc_popen(char *argv[])
{
	result res;

	std::string full_command;
	int i = 0;
	while (true)
	{
		if (argv[i] == nullptr)
		{
			break;
		}
		full_command.append(argv[i]);
		full_command.append(" ");
		i++;
	}
	FILE *cmd_out = popen(full_command.c_str(), "r");
	if (cmd_out == nullptr)
	{
		res.error = -errno;
	}
	else
	{
		char *out = nullptr;
		size_t len = 0;
		ssize_t readbytes;
		while ((readbytes = getline(&out, &len, cmd_out)) >= 0)
		{
			if (out[readbytes - 1] == '\n')
				out[readbytes - 1] = '\0';
			res.output.emplace_back(out);
		}
		free(out);
		int status = pclose(cmd_out);
		if(!WIFEXITED(status) || WEXITSTATUS(status) != 0)
		{
		      res.error = status;
		}
	}

	return res;
}

int main()
{
	const char *argv[] = {
		"docker",
		"--version",
		nullptr
	};

	auto res = runsc((char **)argv);
	printf("run with exit code: %d\n", res.error);
	for (const auto &o : res.output) {
	 	printf("%s\n", o.c_str());
	}
	
	auto res_popen = runsc_popen((char **)argv);
	printf("run_popen with exit code: %d\n", res_popen.error);
	for (const auto &o : res_popen.output) {
	 	printf("%s\n", o.c_str());
	}
	return 0;
}
```

File: userspace/libscap/engine/gvisor/runsc.cpp [hunk 1]
```diff
@@ -30,36 +26,37 @@ namespace runsc {
...
+		i++;
 	}
-
-	pid_t pid = vfork();
-	if(pid > 0) {
-		int status;
-
-		close(pipefds[1]);
-		wait(&status);
+	FILE *cmd_out = popen(full_command.c_str(), "r");
```
Comment (LucaGuerra) [inline userspace/libscap/engine/gvisor/runsc.cpp:40]: is it possible to call `runsc` directly without going through `sh -c`? This is what `popen` does AFAICT. It did before via `execvp`

Comment (FedeDP) [reply userspace/libscap/engine/gvisor/runsc.cpp:40]: Oh i will double check, thanks!

Comment (FedeDP) [reply userspace/libscap/engine/gvisor/runsc.cpp:40]: Oh wait, which `sh -c`? i can't find it :laughing:

Comment (LucaGuerra) [reply userspace/libscap/engine/gvisor/runsc.cpp:40]: https://man7.org/linux/man-pages/man3/popen.3.html

>       The command argument is a pointer to a null-terminated string
>       containing a shell command line.  This command is passed to
>       /bin/sh using the -c flag; interpretation, if any, is performed
>       by the shell.

Comment (FedeDP) [reply userspace/libscap/engine/gvisor/runsc.cpp:40]: Oh thanks, given the man page entry, i don't think it is feasible to skip the `sh -c` for popen.
Is this a problem?

**Comment by FedeDP** [comment]
New function test:
```
#include <sys/wait.h>
#include <ext/stdio_filebuf.h>
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <string>
#include <vector>
#include <sstream>
#include <cstring>

struct result {
   int error = 0;
   std::vector<std::string> output;
};

result runsc(char *argv[])
{
	result res;
	int pipefds[2];

	int ret = pipe(pipefds);
	if(ret)
	{
		return res;
	}

	pid_t pid = vfork();
	if(pid > 0)
	{
		int status;

		close(pipefds[1]);
		wait(&status);
		if(!WIFEXITED(status) || WEXITSTATUS(status) != 0)
		{
			res.error = status;
			return res;
		}

		__gnu_cxx::stdio_filebuf<char> filebuf(pipefds[0], std::ios::in);
		std::string line;
		std::istream is(&filebuf);

		while(std::getline(is, line))
		{
			res.output.emplace_back(std::string(line));
		}
	}
	else
	{
		close(pipefds[0]);
		dup2(pipefds[1], STDOUT_FILENO);
		execvp("docker", argv);
		exit(1);
	}

	return res;
}

result runsc_read(char *argv[])
{
	result res;
	int pipefds[2];

	int ret = pipe(pipefds);
	if(ret)
	{
		return res;
	}

	pid_t pid = vfork();
	if(pid > 0)
	{
		int status;

		close(pipefds[1]);
		wait(&status);
		if(!WIFEXITED(status) || WEXITSTATUS(status) != 0)
		{
			res.error = status;
			return res;
		}


		std::string line;

		char buf[512] = {0};
		std::stringstream ss;
		size_t len;

		// Read until EOF
		while ((len = read(pipefds[0], buf, sizeof(buf) - 1)) != 0)
		{
			ss << buf;
			// Be smart: only reset first len bytes
			memset(buf, 0, len);
		}

		while(std::getline(ss, line))
		{
			res.output.emplace_back(line);
		}
	}
	else
	{
		close(pipefds[0]);
		dup2(pipefds[1], STDOUT_FILENO);
		execvp("docker", argv);
		exit(1);
	}

	return res;
}

int main()
{
	const char *argv[] = {
		"docker",
		"--version",
		nullptr
	};

	auto res = runsc((char **)argv);
	printf("run with exit code: %d\n", res.error);
	for (const auto &o : res.output) {
	 	printf("%s\n", o.c_str());
	}
	
	auto res_read = runsc_read((char **)argv);
	printf("run_read with exit code: %d\n", res_read.error);
	for (const auto &o : res_read.output) {
	 	printf("%s\n", o.c_str());
	}
	return 0;
}
```

File: userspace/libscap/engine/gvisor/runsc.cpp (hunk 1 above)
Comment (FedeDP) [reply userspace/libscap/engine/gvisor/runsc.cpp]: Ok i pushed a version that does not rely upon popen.
See the new example test below: https://github.com/falcosecurity/libs/pull/2037#issuecomment-2363390743

File: userspace/libscap/engine/gvisor/runsc.cpp [hunk 2]
```diff
@@ -48,17 +47,26 @@ result runsc(char *argv[]) {
...
 
-		while(std::getline(is, line)) {
-			res.output.emplace_back(std::string(line));
+		char buf[512] = {0};
+		std::stringstream ss;
+		size_t len;
+
+		// Read until EOF
+		while((len = read(pipefds[0], buf, sizeof(buf) - 1)) != 0) {
+			ss << buf;
```
Comment (LucaGuerra) [inline userspace/libscap/engine/gvisor/runsc.cpp:58]: Don't we need to read only until `len`?

File: userspace/libscap/engine/gvisor/runsc.cpp [hunk 3]
```diff
@@ -48,17 +47,26 @@ result runsc(char *argv[]) {
...
+		}
+
+		while(std::getline(ss, line)) {
+			res.output.emplace_back(line);
 		}
 	} else {
 		close(pipefds[0]);
 		dup2(pipefds[1], STDOUT_FILENO);
-		execvp("runsc", argv);
+		execvp("docker", argv);
```
Comment (LucaGuerra) [inline userspace/libscap/engine/gvisor/runsc.cpp:69]: leftover? :D

File: userspace/libscap/engine/gvisor/runsc.cpp (hunk 1 above)
Comment (LucaGuerra) [reply userspace/libscap/engine/gvisor/runsc.cpp]: Thank you Fede

File: userspace/libscap/engine/gvisor/runsc.cpp (hunk 2 above)
Comment (LucaGuerra) [reply userspace/libscap/engine/gvisor/runsc.cpp:58]: Ah I see, the buffer is always filled with zeroes. I understand.

File: userspace/libscap/engine/gvisor/runsc.cpp (hunk 3 above)
Comment (FedeDP) [reply userspace/libscap/engine/gvisor/runsc.cpp:69]: Uh! ops ahha

Comment (FedeDP) [reply userspace/libscap/engine/gvisor/runsc.cpp:69]: greatest catch ever :rofl:

**Review by LucaGuerra — APPROVED** [review]

**Review by leogr — APPROVED** [review]

**Pull request merged**